iree/compiler/Conversion/LinalgToSPIRV/LegalizeNumWorkgroupsFnPass.cpp

```diff
@@ -0,0 +1,126 @@
+        tieShapeOp.operand().getDefiningOp());
+    if (!phOp) return failure();
+    IntegerAttr operandNumAttr =
+        phOp.getAttrOfType<IntegerAttr>(getOperandResultNumAttrName());
+    if (!operandNumAttr) {
+      return phOp.emitRemark("expected operand_result_num attribute");
+    }
+    FuncOp numWorkgroupsFn = phOp.getParentOfType<FuncOp>();
+    rewriter.replaceOpWithNewOp<Shape::TieShapeOp>(
+        tieShapeOp, phOp,
+        numWorkgroupsFn.getArgument(
+            phOp.getAttrOfType<IntegerAttr>(getOperandResultNumAttrName())
+                .getInt()));
+    return success();
+  }
+};
+
+/// Pattern to remove dead `iree.placeholder` ops. They arent removed since they
+/// are tagged as having `MemoryEffect`.
+struct RemoveDeadPlaceholderOp : OpRewritePattern<IREE::PlaceholderOp> {
+  using OpRewritePattern<IREE::PlaceholderOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(IREE::PlaceholderOp phOp,
+                                PatternRewriter &rewriter) const override {
+    if (phOp.use_empty()) {
+      rewriter.eraseOp(phOp);
+      return success();
+    }
+    return failure();
+  }
+};
+
+/// Pass to legalize the function that computes the number of workgroups to use
+/// for launch to be runnable on the host-side.
+struct LegalizeNumWorkgroupsFnPass
+    : public PassWrapper<LegalizeNumWorkgroupsFnPass, OperationPass<ModuleOp>> {
+  LegalizeNumWorkgroupsFnPass() = default;
+  LegalizeNumWorkgroupsFnPass(const LegalizeNumWorkgroupsFnPass &pass) {}
+  void runOnOperation() override;
+};
+}  // namespace
+
+static void populateLegalizeNumWorkgroupsFnPattern(
+    MLIRContext *context, OwningRewritePatternList &patterns) {
+  patterns.insert<LegalizeTieShapeOp, RemoveDeadPlaceholderOp>(context);
+}
+
+void LegalizeNumWorkgroupsFnPass::runOnOperation() {
+  ModuleOp module = getOperation();
+  auto fns = module.getOps<FuncOp>();
+
+  OwningRewritePatternList patterns;
+  MLIRContext *context = &getContext();
+  populateLegalizeNumWorkgroupsFnPattern(context, patterns);
+
+  SymbolTable symbolTable(module.getOperation());
+  for (FuncOp fn : fns) {
+    if (!isEntryPoint(fn)) continue;
+    auto numWorkgroupsFnAttr =
+        fn.getAttrOfType<SymbolRefAttr>(getNumWorkgroupsFnAttrName());
+    if (!numWorkgroupsFnAttr) continue;
+    StringRef numWorkgroupsFnName = numWorkgroupsFnAttr.getLeafReference();
+    FuncOp numWorkgroupsFn = symbolTable.lookup<FuncOp>(numWorkgroupsFnName);
+    if (!numWorkgroupsFn) {
+      fn.emitError("unable to find function to compute number of workgroups ")
+          << numWorkgroupsFnName;
+      return signalPassFailure();
+    }
+    if (failed(applyPatternsAndFoldGreedily(numWorkgroupsFn, patterns)))
+      return signalPassFailure();
+  }
+}
+
+std::unique_ptr<OperationPass<ModuleOp>> createLegalizeNumWorkgroupsFnPass() {
+  return std::make_unique<LegalizeNumWorkgroupsFnPass>();
+}
+
+static PassRegistration<LegalizeNumWorkgroupsFnPass> pass(
```

Comment:
Instead of init/legalize, I'd think declare/define would be better. But your choice.